Add virtual environment (venv) tag (#1702)
python-discord/bot: bot/resources/tags/venv.md

**Virtual Environments**

Virtual environments are isolated Python environments, which make it easier to keep your system clean and manage dependencies. By default, when activated, only libraries and scripts installed in the virtual environment are accessible.

> To create a new virtual environment, you can use the `venv` module: `python -m venv .venv`

Reviewer: ```suggestion
To create a new virtual environment, you can use the standard library `venv` module: `python3 -m venv .venv` (replace `python3` with `python` or `py` on windows)
```
May need `python3` on unix, so probably worth changing to something like this. Also mention that's it's in the standard library

Then, to activate the new virtual environment:

PowerShell: `.venv\Scripts\Activate.ps1`

> Bash: `source .venv/bin/activate`

Reviewer: I think it would also be worth adding the command for cmd,  `.venv\Scripts\activate.bat`

> Note: On Windows, you may need to change the [execution policy](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_execution_policies) first:

Reviewer: This sort of feels a little out of scope for the tag. Would be nice if we could just link to this section in the venv docs next to the command above, although i'm not sure you can specifically link to that part. You could try linking to another page, like https://dev.to/aka_anoop/enabling-virtualenv-in-windows-powershell-ka3, may be nicer to have this info on our site at some point though,

Author: The issue with those pages is that they suggest changing the execution policy to `Unrestricted`, whereas `RemoteSigned` would be preferred -- as the Python docs suggest.

Until we get a page for the site, keeping something about execution policies here will probably be useful.

Reviewer: Another possible option could be to link to a gist, something like https://gist.github.com/wookie184/4bec5f2785ad72736a6a46bb4cc7defc#file-executionpolicy-md, although if you'd like to leave the note in the tag for now I think that's fine too.

Reviewer 2: I don't think length is a problem, but I do think the prominence is a problem. Someone reading the tag might think that this is immediately important/applies all the time, given the font-size of codeblocks and the fact it comes before the very end of the tag, and this isn't strictly true, you only need to run the command once. It takes a lot of mindshare and I think that can be improved. 

Try moving it to the very, very end of the tag, as well as sticking it within \`regular code tags\` rather than a \`\`\`code block\`\`\`, and see if that's better?

Author: My reasoning for its placement is that it follows directly after the activation examples, which is the stage where changing the execution policy is required. I'll make some changes to suggest that it's only required once.

I agree that it is rather prominent, however. Changing to single backticks seems to work, but it should probably be on a new line to avoid wrapping.

```powershell
Set-ExecutionPolicy -ExecutionPolicy RemoteSigned -Scope CurrentUser
```
Packages can then be installed to the virtual environment using `pip`, as normal.

> For more information, take a read of the [documentation](https://docs.python.org/3/library/venv.html).

Reviewer: ```suggestion
For more information, take a read of the [documentation](https://docs.python.org/3/library/venv.html). Your editor may also have venv integration, in which case you should check out your editors documentation. Here's the docs for [VSCode](https://code.visualstudio.com/docs/python/environments#_select-and-activate-an-environment) and [PyCharm](https://www.jetbrains.com/help/pycharm/creating-virtual-environment.html)

There are also tools such as [poetry](https://python-poetry.org/docs/basic-usage/) and [pipenv](https://pipenv.pypa.io/en/latest/) you can use for managing creation of venvs as well as project dependency management, among other things
```

**Virtual Environments**

Virtual environments are isolated Python environments, which make it easier to keep your system clean and manage dependencies. By default, when activated, only libraries and scripts installed in the virtual environment are accessible, allowing dependency conflicts to be resolved, and your program requirements to be found easily.

To create a new virtual environment, you can use the standard library `venv` module: `python3 -m venv .venv` (replace `python3` with `python` or `py` on Windows)

Then, to activate the new virtual environment:

**Windows** (PowerShell): `.venv\Scripts\Activate.ps1`
or (Command Prompt): `.venv\Scripts\activate.bat`
**MacOS / Linux** (Bash): `source .venv/bin/activate`

> Note: On Windows, you may need to change the [execution policy](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_execution_policies) first. This is only required once:

Reviewer: ```suggestion
Note: When using Windows PowerShell, you may need to change the [execution policy](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_execution_policies) first. This is only required once:
```
This doesn't apply to cmd, so we should specify it's PowerShell only. I also like bast's suggestion of moving this to the bottom

> For more information, take a read of the [documentation](https://docs.python.org/3/library/venv.html). Your IDE, for example [VSCode](https://code.visualstudio.com/docs/python/environments#_select-and-activate-an-environment) or [PyCharm](https://www.jetbrains.com/help/pycharm/creating-virtual-environment.html), likely has `venv` integration.

Reviewer: I think the bit about other IDEs should be reworded. I think it should be more clear that the links relate to the corresponding editors' virtual environment documentation rather than just information about the editors itself. I also think it should be specifically recommending looking at your editors venv documentation.

For example, it is common that a user may set up a venv using a guide similar to this, and then be confused when they click the green run button in the IDE and it runs with the system python rather than in the venv.

Something like:
```
If you run your code through your editor you should check it's docs so you can set it up to use your venv, for example, see the [VSCode](link) or [Pycharm](link) docs.
```
Thoughts?

Author: Yep, I struggled to keep that section short, while keeping it relevant. I think that sums it up nicely.

Just a slight reword:

> If you run code through your editor, check the relevant documentation on how to use your new virtual environment. For example, see the [VSCode](https://code.visualstudio.com/docs/python/environments#_select-and-activate-an-environment) or [PyCharm](https://www.jetbrains.com/help/pycharm/creating-virtual-environment.html) docs.

Is that alright?

Reviewer: Yeah, that sounds great.

This is maybe slightly more concise again

```
If you run code through your editor, check it's documentation on how to make it use your virtual environment. For example, see the VSCode or PyCharm docs.
```

Anything along these lines or what you suggested is good though

**Virtual Environments**

> Virtual environments are isolated Python environments, which make it easier to keep your system clean and manage dependencies. By default, when activated, only libraries and scripts installed in the virtual environment are accessible.

Reviewer 2: I think we can fit a sentence in here explaining why clean dependencies are important. We have an opportunity to make this more beginner friendly without adding too much, and I think we should take it. Eg:
 - knowing what your program actually requires
 - dependency conflicts
 - partial upgrades rather than having to update every script on your system at once

Author: Okey doke, totally agree. I'll try to throw some of those in there.

To create a new virtual environment, you can use the standard library `venv` module: `python3 -m venv .venv` (replace `python3` with `python` or `py` on Windows)

Then, to activate the new virtual environment:

PowerShell: `.venv\Scripts\Activate.ps1`
Command Prompt: `.venv\Scripts\activate.bat`

> Bash: `source .venv/bin/activate`

Reviewer 2: We should note that bash applies to linux/osx. It took me a moment ( despite extensive familiarity with all three here and the os's and python ) to realize that yeah, the line marked "bash" was the one that I was looking for. If we put (OSX/Linux) in there, or perhaps something better if you can figure it out, that might help make things clearer and more obvious.

Author: My concern with specifying an OS would be that bash, for example, could be for Linux, OSX or WSL. Likewise, PowerShell can be installed on platforms other than Windows. The only specific difference between the three is the shell itself.

I guess this is a little pedantic, and can probably be ignored. Thoughts?

Reviewer 2: WSL is linux, technically. While you can get bash underneath windows proper with MingW, the reason to mention the OS is to catch the people who are using it (starting anew on mac, or just getting their feet wet on linux), who think of "terminal" rather than "bash". 

It is a significant portion of the people likely to view the tag, rather than those who've installed powershell on something other than windows, who are unlikely to get confused on which applies since they've chosen to install the tech by name manually. 

If we did have a linux distro of some popularity utilizing powershell as the default we'd want to note that here. There isn't one, so thankfully we won't be tempted to.

Author: > If we did have a linux distro of some popularity utilizing powershell as the default

Sent shudders down my spine haha

Yeah, with the target audience, it's probably best to choose clarity over accuracy. Would something like below be okay?
> Shell (OS): `path/to/activate`

Reviewer 2: Well, I'm not sure what the best option is, but again, we're painting targets. Like how bow targets are concentric rings and the comment button here is green and on the bottom right, we want the people searching the tag for what activate command they need to have their eyes immediately go to the right spot. 

I do think duplicating Windows (powershell) might be too much.. but maybe not. I'll defer to other's judgement on that.

Author: Writing "Windows" twice I think would be redundant. I've used bold for the OS to draw attention to the determining factor. This is the best way I can think to distinguish between PowerShell and Command Prompt:

**Windows** (PowerShell): `.venv\Scripts\Activate.ps1`
or (Command Prompt): `.venv\Scripts\activate.bat`
**MacOS / Linux** (Bash): `source .venv/bin/activate`

Reviewer 3: I think that sounds pretty good and clear.

Note: On Windows, you may need to change the [execution policy](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_execution_policies) first:
```powershell
Set-ExecutionPolicy -ExecutionPolicy RemoteSigned -Scope CurrentUser
```
Packages can then be installed to the virtual environment using `pip`, as normal.

> For more information, take a read of the [documentation](https://docs.python.org/3/library/venv.html). Your IDE, for example [VSCode](https://code.visualstudio.com/docs/python/environments#_select-and-activate-an-environment) or [PyCharm](https://www.jetbrains.com/help/pycharm/creating-virtual-environment.html), may also have `venv` integration.

Reviewer 2: Would changing `may` to `is likely to` or `likely has` make this better?

**Virtual Environments**

> Virtual environments are isolated Python environments, which make it easier to keep your system clean and manage dependencies. By default, when activated, only libraries and scripts installed in the virtual environment are accessible, allowing dependency conflicts to be resolved, and your program requirements to be found easily.

Reviewer 2: I think the wording here could be a little better. Perhaps `preventing cross-project dependency conflicts and easy isolation of requirements`?

Author: That's a lot cleaner - done :+1:

To create a new virtual environment, you can use the standard library `venv` module: `python3 -m venv .venv` (replace `python3` with `python` or `py` on Windows)

Then, to activate the new virtual environment:

**Windows** (PowerShell): `.venv\Scripts\Activate.ps1`
or (Command Prompt): `.venv\Scripts\activate.bat`
**MacOS / Linux** (Bash): `source .venv/bin/activate`

Note: On Windows, you may need to change the [execution policy](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_execution_policies) first. This is only required once:

> `Set-ExecutionPolicy -ExecutionPolicy RemoteSigned -Scope CurrentUser`

Reviewer 2: I really do think this should be at the very bottom. As it is it somewhat interrupts the flow of the text like an overly large () in the middle of a paragraph.
